common/lanelet2_extension/lib/CarmaTrafficLight.cpp

```diff
@@ -0,0 +1,137 @@
+#include "RegulatoryHelpers.h"
+
+namespace lanelet
+{
+// C++ 14 vs 17 parameter export
+#if __cplusplus < 201703L
+constexpr char CarmaTrafficLight::RuleName[];  // instantiate string in cpp file
+#endif
+
+ConstLineStrings3d CarmaTrafficLight::stopLine() const
+{
+  return getParameters<ConstLineString3d>(RoleName::RefLine);
+}
+
+LineStrings3d CarmaTrafficLight::stopLine()
+{
+  return getParameters<LineString3d>(RoleName::RefLine);
+}
+
+CarmaTrafficLight::CarmaTrafficLight(const lanelet::RegulatoryElementDataPtr& data) : RegulatoryElement(data)
+{}
+
+std::unique_ptr<lanelet::RegulatoryElementData> CarmaTrafficLight::buildData(Id id, LineStrings3d stop_line, Lanelets lanelets,
+                                                                Areas areas)
+{
+  for (auto ls : stop_line)
+  {
+    if (ls.empty()) throw lanelet::InvalidInputError("Empty linestring was passed into CarmaTrafficLight buildData function");
+  }
+  
+  // Add parameters
+  RuleParameterMap rules;
+
+  rules[lanelet::RoleNameString::RefLine].insert(rules[lanelet::RoleNameString::RefLine].end(), stop_line.begin(),
+                                                 stop_line.end());
+
+  // Add attributes
+  AttributeMap attribute_map({
+      { AttributeNamesString::Type, AttributeValueString::RegulatoryElement },
+      { AttributeNamesString::Subtype, RuleName },
+  });
+
+  return std::make_unique<RegulatoryElementData>(id, rules, attribute_map);
+}
+
+CarmaTrafficLightState CarmaTrafficLight::getState()
+{
+  return predictState(ros::Time::now());
+}
+
+CarmaTrafficLightState CarmaTrafficLight::predictState(ros::Time time_stamp)
+{
+  if (recorded_time_stamps.empty())
+  {
+    throw::lanelet::InvalidObjectStateError("CarmaTrafficLight doesn't have any recorded states of traffic lights");
+  }
+  if (recorded_time_stamps.size() == 1) // if only 1 timestamp recorded, this signal doesn't change
+  {
+    return recorded_time_stamps.front().second;
+  }
+  // shift starting time to the future or to the past to fit input into a valid cycle
+  ros::Duration accumulated_offset_duration;
+  double offset_duration_dir = recorded_time_stamps.front().first > time_stamp ? -1.0 : 1.0; // -1 if past, +1 if time_stamp is in future
+
+  if (offset_duration_dir < 0) 
+  {
+    while (recorded_time_stamps.front().first - accumulated_offset_duration > time_stamp)
+    {
+      accumulated_offset_duration += fixed_cycle_duration;
+    }
+  }
+  else
+  {
+    while (recorded_time_stamps.front().first + accumulated_offset_duration < time_stamp)
+    {
+      accumulated_offset_duration += fixed_cycle_duration;
+    }
+  }
+  
+  // iterate through states in the cycle to get the signal
+  for (int i = 0; i < recorded_time_stamps.size(); i++)
+  {
+    if (recorded_time_stamps[i].first.toSec() + offset_duration_dir * accumulated_offset_duration.toSec() >= time_stamp.toSec())
+    { 
+      return recorded_time_stamps[i].second;
```

Comment:
Is the recorded_time_stamps the start time or end time of the phase? If its the start time, then this would need to be i-1 right?

Reply:
https://github.com/usdot-fhwa-stol/carma-msgs/blob/develop/cav_msgs/msg/TimeChangeDetails.msg#L51

float32 min_end_time

which is the only filed that is required in the msg

Reply:
we are using end_time